optimize/types.go

```diff
@@ -129,11 +129,9 @@ type Problem struct {
 	Grad func(grad []float64, x []float64) []float64
 
 	// Hess evaluates the Hessian at x and stores the result in-place in hess.
-	// Hess must not modify x. Hess may use (and return) the provided Symmetric
-	// if it is non-nil, or must allocate a new Symmetric otherwise. Minimize
-	// will 'give back' the returned Symmetric where possible, allowing Hess
-	// to use a type assertion on the provided matrix.
-	Hess func(hess mat.Symmetric, x []float64) mat.Symmetric
+	// Hess must not modify x. Hess must use the provided mat.SymDense, and
+	// must resize it if it is zero-sized.
+	Hess func(hess *mat.SymDense, x []float64)
```

Comment:
I'm slightly concerned about the discord between `Grad` and `Hess` but I don't have an alternative other than Grad taking a non-nil pointer to correctly sized or zero sized slice.

Reply:
I've been thinking about that. The difference and justification is that `Hess` takes a pointer value destination and `Grad` takes a non-pointer destination. This means that in order to resize the `Grad` destination in a way that the caller can see, there needs to be a returned value. This doesn't need to happen with the `Hess` destination. We could have them always returning the destination, but then `Hess` has two potential sources of truth, the passed in parameter and the returned value. Since these are potentially user-implemented functions, this is a crack for bugs to live in.

In short, I'm not wildly happy with it, but I think it's the least worst.

Reply:
If we are doing it this way, we could go back to the old `optimize` way, where the user didn't need to care about it (and had to store in-place). We do need the ability for the `Method` to set the value to `nil`, but that case could checked and allocated in `Minimize`.  This was not possible for an arbitrary `Symmetric` but is very easy with a fixed `*SymDense`.

Reply:
This would be simpler for users and would still mesh with `fd`, since the matrix would always be allocated.

Reply:
I'm unclear what the old `optimize` way was. What was the signature?

Reply:
```
Grad: func(dst, x []float64)
Hess: func(dst *mat.SymDense, x []float64)
```

The difference being that `Minimize` guarantees that `len(dst) == len(x)` in the call to grad, and `dst.Symmetric() == len(x)` in the call to `Hess`. In other words, `Minimize` guarantees they are the right size and the user doesn't have to worry about checking if they are nil or not.

Reply:
Ah. OK. Would you like me to update this for that behaviour?

Reply:
I think so, yes. That check should probably happen in `evaluate`, minimize.go ~ line 470 (and there would be corresponding documentation changes in `Problem`)

Reply:
There's a TODO for consideration. I can do that in this PR if you want; whether the allocation of the gradient slice is when it is nil, or when it is zero-length.

Reply:
We can leave as a TODO. I suspect the correct behavior is to only check if the relevant entry is `nil`. These values are never allocated by the provider of `Problem`, this is only for communicating between `Minimize` and `Method`. The reason to allow nil values is so the `Method` can safely save the values. It seems like allowing zero values as well could be ripe for bugs, in that it's mid-way between explicitly keeping the data and explicitly destroying it.

Reply:
That would suggest that the Hessian should also only be a nil check then.

Reply:
Done

Reply:
This is better.